### PR TITLE
Made sorting on NdMapping optional

### PR DESCRIPTION
**Pull request by philippjfr**
Let's users provide an explicitly sorted NdMapping. Still need to add a warning when a regular dict and sort=False are supplied since sort order is nondeterministic in that case.

~~Edit: Widgets are currently assuming sorted keys and NdElement seems to have been sorting in some cases.~~

Addresses: https://github.com/ioam/holoviews/issues/1216

Fixes: https://github.com/ioam/holoviews/issues/1042

### CODE REVIEW

**Comment by philippjfr** [comment]
Ready to review/merge when tests pass.

File: holoviews/core/ndmapping.py [hunk 1]
```diff
@@ -49,11 +49,11 @@ def __init__(self, enabled):
         self.enabled = enabled
 
     def __enter__(self):
```
Comment (jlstevens) [inline holoviews/core/ndmapping.py]: The docstring of this context_manager should be updated. As now ``sort=False`` is valid, it should just say it disables sorting regardless of whether the NdMapping has ``sort=True`` or ``sort=False``. 

I also think the line 'Should only be used if values are guaranteed to be sorted before or after the operation is performed.' should just say something else - maybe just that the initial ordering (whatever it is) should be preserved?

File: holoviews/core/ndmapping.py [hunk 2]
```diff
@@ -89,20 +89,17 @@ class MultiDimensionalMapping(Dimensioned):
...
-            map_type = type(initial_items)
-            own_params = self.params()
-            new_params = dict(initial_items.get_param_values(onlychanged=True))
-            if new_params.get('group') == map_type.__name__:
-                new_params.pop('group')
-            params = dict({name: value for name, value in new_params.items()
-                           if name in own_params}, **params)
+        if isinstance(initial_items, MultiDimensionalMapping):
+            params = dict(util.get_param_values(initial_items),
+                          **dict({'sort': self.sort}, **params))
```
Comment (jlstevens) [inline holoviews/core/ndmapping.py]: Nice to see the simplified constructor!

File: holoviews/core/ndmapping.py [hunk 3]
```diff
@@ -169,11 +170,7 @@ def _add_item(self, dim_vals, data, sort=True, update=True):
 
         for dim, val in valid_vals:
             vals = self._cached_index_values[dim.name]
-            if vals == 'initial': self._cached_index_values[dim.name] = []
-            if not self._instantiated and self.get_dimension(dim).values == 'initial':
-                if val not in vals:
-                    self._cached_index_values[dim.name].append(val)
-            elif vals and val is not None and val not in vals:
```
Comment (jlstevens) [inline holoviews/core/ndmapping.py]: Was this chunk of logic made redundant by some other change or was it always redundant?

File: holoviews/core/ndmapping.py [hunk 4]
```diff
@@ -283,10 +279,9 @@ def groupby(self, dimensions, container_type=None, group_type=None, **kwargs):
         container_type = container_type if container_type else type(self)
         group_type = group_type if group_type else type(self)
         dimensions = [self.get_dimension(d, strict=True) for d in dimensions]
-        sort = not self._sorted
         with item_check(False):
             return util.ndmapping_groupby(self, dimensions, container_type,
-                                          group_type, sort=sort, **kwargs)
+                                          group_type, sort=True, **kwargs)
```
Comment (jlstevens) [inline holoviews/core/ndmapping.py]: Probably worth mentioning either in a comment - or maybe even the docstring - that sorting will always be applied.

File: holoviews/core/ndmapping.py (hunk 2 above)
Comment (philippjfr) [inline holoviews/core/ndmapping.py]: That's where the bug reported in #1042 was, this was written very early on before we had utilities for this kind of thing.

**Comment by jlstevens** [comment]
Looks good. Most of my comments relate to documentation issues...

I agree a ``ValueError`` is better than a warning when ``sort=False`` and an unsortable dictionary is supplied. It is probably best to completely disallow the nondeterministic case rather than simply warning.

File: holoviews/core/ndmapping.py (hunk 3 above)
Comment (philippjfr) [inline holoviews/core/ndmapping.py]: This was my early hacky approach to define a fixed order, I think it has been broken for a long time.

File: holoviews/core/ndmapping.py (hunk 4 above)
Comment (philippjfr) [inline holoviews/core/ndmapping.py]: Will do.

Comment (philippjfr) [inline holoviews/core/ndmapping.py]: Added to the docstring above.

File: holoviews/core/ndmapping.py (hunk 1 above)
Comment (philippjfr) [inline holoviews/core/ndmapping.py]: Docstring updated.

**Comment by jlstevens** [comment]
Tests are now fixed (pr build is passing at least) and you've made the docstring changes I suggested. Merging!

**Pull request merged**